node/toolrunner.ts

```diff
@@ -580,200 +580,234 @@ export class ToolRunner extends events.EventEmitter {
-                cp.stdin.end();
-                if(waitingEvents == 0) {
-                    if (error) {
-                        defer.reject(error);
-                    } else {
-                        defer.resolve(returnCode);
-                    }
+
+                cp2.stdin.end();
+                state.CheckComplete();
+            });
+
+            var stdbuffer: string = '';
+            cp2.stdout.on('data', (data: Buffer) => {
+                this.emit('stdout', data);
+
+                if (!options.silent) {
+                    options.outStream.write(data);
                 }
+
+                this._processLineBuffer(data, stdbuffer, (line: string) => {
+                    this.emit('stdline', line);
+                });
             });
 
-        } else {
-            waitingEvents++;
-            cp = child.spawn(this._getSpawnFileName(), this._getSpawnArgs(options), this._getSpawnOptions(options));
-        }
+            var errbuffer: string = '';
+            cp2.stderr.on('data', (data: Buffer) => {
+                state.process2Stderr = true;
+                this.emit('stderr', data);
 
-        var stdbuffer: string = '';
-        cp.stdout.on('data', (data: Buffer) => {
-            this.emit('stdout', data);
+                if (!options.silent) {
+                    var s = options.failOnStdErr ? options.errStream : options.outStream;
+                    s.write(data);
+                }
 
-            if (!options.silent) {
-                options.outStream.write(data);
-            }
+                this._processLineBuffer(data, errbuffer, (line: string) => {
+                    this.emit('errline', line);
+                });
+            });
 
-            this._processLineBuffer(data, stdbuffer, (line: string) => {
-                this.emit('stdline', line);
+            cp2.on('error', (err) => {
+                state.process2Error = err.message;
+                state.process2Exited = true;
+                state.process2Closed = true;
+                state.CheckComplete();
             });
-        });
 
-        var errbuffer: string = '';
-        cp.stderr.on('data', (data: Buffer) => {
-            this.emit('stderr', data);
+            cp2.on('exit', (code, signal) => {
+                state.process2ExitCode = code;
+                state.process2Exited = true;
+                this._debug(`Exit code ${code} received from tool '${this.pipeOutputToTool.toolPath}'`);
+                state.CheckComplete()
+            });
 
-            success = !options.failOnStdErr;
-            if (!options.silent) {
-                var s = options.failOnStdErr ? options.errStream : options.outStream;
-                s.write(data);
-            }
+            cp2.on('close', (code, signal) => {
+                state.process2ExitCode = code;
+                state.process2Closed = true;
+                this._debug(`STDIO streams have closed for tool '${this.pipeOutputToTool.toolPath}'`)
```

Comment:
do we also want to set state.processExited = true; in case we miss the exit event.